IQSS/dataverse
# avoid expensive Solr join for public dvObjects in search (experimental)

**What this PR does / why we need it**:

We were advised yesterday by a Solr expert that we should avoid Solr joins whenever possible because they can impact performance.

This pull requests removes the join used when Guest users search. However, dvObjects that would have been found by IP Groups will no longer be found. In addition, a full reindex is required as the new field, a dynamic field of boolean type called `publicObject_b`, is located on the content documents themselves (dvObjects) rather than permission documents (which we join with).

[**Edit** from Leonid:] The scope of the PR has been expanded to cover ALL the searches - not just for Guest users - and always rely on the newly-added boolean for selecting publicly-available objects. When the user initiating the search, whether a Guest or an AuthenticatedUser, is also a member of one or more groups (remember, Guests *can* be parts of some groups, such as IP groups), a join is added that is limited in scope to just these restricted groups. I was able to figure out a query that works on a *combination* of this boolean `publicObject_b:true` AND an additional limited join. See under "How to test" on how to observe the effect of these changes.

**Which issue(s) this PR closes**:

- Closes #10554

This is related to the ongoing Solr perf effort:

- #10469

**Special notes for your reviewer**:

For simplicity, I implemented a single experimental feature flag but we can certainly change this to multiple settings: one to tell Dataverse to start indexing content documents with `publicObject_b` and another setting to switch search over to start using them (and avoid the join). [**Edit** from Leonid: this has been done, there are now 2 separate feature flags]

I did very light testing on this (and didn't test IP Groups at all). I did ensure that SearchIT passed on my machine.

**Suggestions on how to test this**:

After enabling the feature:

- Make sure guest users can still find expected dvObjects.

[**Edit** from Leonid:] The QA needs to involve confirming that searches are still working exactly the same as before, *without* these new experimental feature flags enabled. I'm fairly confident that the code handling that has been left unchanged; but it had to be at least moved around a little, so should be prudent to confirm. 
In order to see the full impact of the new optimization in searches, the PR must be tested on the perf. test system, or a similar-size instance. 
The "before" part, confirming the scale of performance degradation on an instance the size of ours, with the current implementation: 
- Create a local account in perf (assuming you don't have a local account in prod. yet; note that HarvardKey logins are not going to work there);
- Deploy develop branch on the perf. cluster; 
- Load some collection pages, as a guest, super- and regular logged in users. Confirm that the performance is somewhat tolerable, on an instance that is NOT under heavy load. 
- Now start a full reindex (```http://localhost:8080/api/admin/index```). Watch the load time for collection pages become abysmal, for both the guest an a logged in non-superuser. Note that for the superuser, ```dataverseAdmin``` the performance will still be somewhat ok - because no permission joins are performed for such. 

The "after" part: 
- Deploy this PR build, start the reindex again. Watch the difference in performance, for both the guest, and logged in regular users. 
Notes: 
- A full reindex has already been performed on the perf. system with the new feature enabled; i.e., all the public objects already have the new boolean flag ```publicObject:true``` in them. 
- Note that the database is a copy of prod., and therefore has collection creation similarly disabled for non-superusers. So if you need a collection that's only visible to your local user, for testing, the only way is to create the collection as the ```dataverseAdmin``` user, then make your local user an admin (under "Permissions"), who can then proceed creating datasets etc. in it. 
- Both feature flags are already enabled as JVM options in the ```domain.xml```. 


**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

No.

**Is there a release notes update needed for this change?**:

Yes, included.

**Additional documentation**:

Included.
